### ISSUE TITLE
Can't set a custom ouput device on SoundCloud

### ISSUE THREAD

**Issue by mateo-m** [ish-/AuRo#7]
I tried setting a custom output device on [SoundCloud](https://soundcloud.com) but the extension seems to break whenever I try to play a music on a different device than the default device.

<ins>Here's a step-by-step recap of messages sent in the console</ins>:
1. Load the website: ![1 - default audio device](https://user-images.githubusercontent.com/23667794/191968248-7e5a98b4-c139-4db5-a91e-c285184eecf1.png)
2. Set a custom device (**without** saving domain): ![2 - set custom device (without saving domain)](https://user-images.githubusercontent.com/23667794/191968313-cd81a724-1f61-419c-8d3d-18586d600cb5.png)
3. Set custom audio device (**with** saving domain): ![3 - set custom device (with saving domain)](https://user-images.githubusercontent.com/23667794/191968425-b59cdab4-5733-4db8-9895-5fd3b200d567.png)
4. Load website with saved domain for custom audio device: ![4 1 - load soundcloud with saved domain](https://user-images.githubusercontent.com/23667794/191969064-f7b7a2ad-7347-44d0-a743-b4fa6d8ae4b4.png)
5. Start music with saved domain for custom audio device: ![4.2 - start music with saved domain](https://user-images.githubusercontent.com/23667794/191968706-f1725fb5-2ab2-45b4-95ee-01e6ac71ea1f.png)

Does anybody have a clue on why this is happening? @ish- 

<ins>P.S.</ins>: the extension works perfectly on YouTube.

**Comment by mateo-m** [comment]
@kawazoe, I saw that you have a `v3` branch where you made a bunch of changes to support the v3 of the manifest file.
Did you notice similar to mine on your branch's latest build ?

**Comment by kawazoe** [comment]
@mateo-m I'm looking into it right now.

Soundcloud sets a policy on some domains to deny the use the the microphone (which is required to enumerate the devices) at the HTTP level via https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy, so the browser won't even request permission to the user. It turns out that one of these domains is secure.soundcloud.com, which is included as an iframe into every soundcloud page. That seems to trip the extension since both the main frame and the secure frame tries to reply with a list of medias, and the secure frame always answers first.

I'm trying to figure out a way to make sure only the main frame gets to enumerate devices. I'll get back to you when I find a fix.

**Comment by kawazoe** [comment]
@mateo-m Quick update. I managed to fix the permission issue that I was having, and now I get the same error as you. After some research, I still cannot figure out what causes it. It's an error that happens when the browser fails to set the output device. Right now, I'm looking into what makes SoundCloud different from other players like YouTube or Tidal. Still no luck...

**Comment by kawazoe** [comment]
@mateo-m Ok so I have a final answer. SoundCloud uses AudioContext to reconstruct the audio from multiple chunks before sending them to an HTMLAudioElement. AudioContext does not yet support (though it has been proposed recently) setSinkId. Until browsers implements this new spec, there is no way to change the output device on SoundCloud.

**Comment by mateo-m** [comment]
> @mateo-m SoundCloud uses AudioContext to reconstruct the audio from multiple chunks before sending them to an HTMLAudioElement. AudioContext does not yet support (though it has been proposed recently) setSinkId. Until browsers implements this new spec, there is no way to change the output device on SoundCloud.

@kawazoe Okay thank you for all this information, didn't know SoundCloud handled how their music is played differently from every other website (thinking about YT and Spotify).

**Comment by mateo-m** [comment]
Something kind of unrelated to my original issue, but related to you `v3` branch @kawazoe:
First of all, I love the work you've done, switching to manifest v3, refactoring the code to use modern JS, great stuff!

I just wanted to let you know that aftert installing packages and building the extension from your `v3` branch, the extension crashes with this error:
![screenshot1](https://user-images.githubusercontent.com/23667794/207065189-92538c40-7b59-4278-9525-4852fa4269ab.png)
![screenshot2](https://user-images.githubusercontent.com/23667794/207065203-d26b744f-eb64-4027-98d0-8daf31b04308.png)

**Comment by ish-** [comment]
@mateo-m cant approve that

**Comment by kawazoe** [comment]
@mateo-m I'm not too sure of how you got this error. I did find an issue when the extension is triggered on system tabs, but I don't think this is the one you are experiencing.

Can you give me some repro steps? Try to include how you built the extension and added it to your browser as we might do things differently.

I also pushed an update to the readme and build process to make it easier to debug.

**Comment by mateo-m** [comment]
<ins>Steps to reproduce the issue I reported in my [previous comment](https://github.com/ish-/AuRo/issues/7#issuecomment-1346567594)</ins>:
1. Clone `git@github.com:kawazoe/AuRo.git` — `git clone git@github.com:kawazoe/AuRo.git`
2. CD into the cloned repository — `cd AuRo`
3. Checkout to the `v3` branch — `git checkout v3`
4. Install dependencies — `npm i`
5. Build the extension — `npm run build`
6. Load the built extension as unpacked from the "Extensions" tab
7. An `Errors` button will be visible as soon as the unpacked extension is added to the browser, click said button to get the error I previously reported

**Comment by kawazoe** [comment]
@mateo-m I am legitimately surprised by this. For me, it seems like uninstalling the extension and reinstalling it isn't enough to trigger your issue.

Have you tried the latest version that I pushed on `v3` three days ago? Could you try it with `npm run watch`? It should enable verbose logging which might help me figure out what's going on.

Also, which browser are you using and which version? Plain old Chrome? Some Chromium variant?

**Comment by kawazoe** [comment]
@mateo-m I tested the build on a Windows machine and my hunch was correct. There's a bug in how rollup-plugin-chrome-extension handles paths on Windows. It isn't updated anymore and has been super-seeded by a vite plugin. I wanted to keep the build process simple so I stuck to rollup anyway.

TLDR, I've applied a small workaround and it should build correctly on Windows now.

**Comment by mateo-m** [comment]
@kawazoe I'm going to test your workaround right away !
And yes, I'm a Windows and Brave user, I should've told you sooner 😅...

**Comment by mateo-m** [comment]
The forked version from the `v3` seems to work flawlessly with your workaround.
Thank you for finding the issue and fixing it this fast !
